### DISCUSSION TITLE
[Template Fitting] Is it ok to have large amount of empty bins in the template?

### DISCUSSION THREAD

**Discussion by zjkjsd** [scikit-hep/iminuit#825] [Q&A] [answered]
Aloha iminuit community and experts,

I am a student working on R(D) and R(D*) measurements at Belle II. The current plan of signal extraction is through a 2d template fitting ( $|\vec{p_D}| + |\vec{p_\ell}|$  vs.  $M_{miss}^2$ ). In my practice using iminuit template fitting, templates are constructed for 6 different signal-like B meson decay modes from signal MC.

![download](https://user-images.githubusercontent.com/60195377/212804683-949a41bb-7f2b-47a5-9ceb-63c093089df4.png)
 
These 2d histograms are flattened and then used as templates. However, as seen in the picture, the 2 fitting quantities ( $|\vec{p_D}| + |\vec{p_\ell}|$  vs.  $M_{miss}^2$ ) are correlated and there is a large number of empty bins in each template. I thought these empty bins might introduce unnecessary degrees of freedom if the templates are used as is. It might be better if these empty bins can be removed before using them as templates. 

So I tested this idea by fitting templates to an independently produced MC sample (test sample). My code snippet and fitting results with and without empty bins are attached below. The empty bins are defined in the test sample.

```
from iminuit import Minuit
from iminuit.cost import poisson_chi2, Template
from IPython.display import display

xedge = np.linspace(0, len(data), len(data)+1)
init_yields = [12000]*6
c4 = Template(data, xedge, temp, method="da")
m4 = Minuit(c4, *init_yields, name=model.config.samples)
m4.limits = (0, None)
m4.migrad()
m4.hesse()
m4.minos()
```
![Screen Shot 2023-01-16 at 6 10 28 PM](https://user-images.githubusercontent.com/60195377/212808792-056b1996-c33b-451f-9856-0800083cc582.png)

Above is the result with empty bins and below is without. The reduced $\chi^2$ changed from 0.4 to 1.2 as expected but the uncertainties increased quite a lot. Now I am not sure if the empty bins should be removed or not. Could you share some insights? Any comments or suggestions are much appreciated!

![Screen Shot 2023-01-16 at 6 10 57 PM](https://user-images.githubusercontent.com/60195377/212809541-a41e3364-613d-4f4a-8966-579d0226bb66.png)

(ps. I don't know why the `Name` column in the 2nd block has a lot of extra spaces...)
Mahalo,
Boyang

**Comment by HDembinski** [comment, marked as answer]
tl;dr: You should keep the empty bins and it makes sense that the fit with empty bins included gives you smaller uncertaintes.

Empty bins in the data are completely ok, no need to cut them. Problematic are empty bins in the templates, more precisely, situations in which **all** templates have zero entries in some bin but the data bin is non-zero. Such bins have to be discarded in the likelihood function, because it is not possible to draw any information from them (all predictions happen to be empty, so there is no way to estimate the amplitudes).

If at least one template is filled for a given bin and the data bin is empty, the Poisson-based template fit can draw information from that. For the Poisson distribution we can compute the probability to observe nothing for a given expected value. This is why your fit with empty bins included gives you more precise estimates for the yields.

**Reply by zjkjsd** [reply]
Thank you very much for your answer! It makes a lot of sense to me! I forgot to mention that kinematic limits prohibit certain regions in the 2d plane, so I need to remove those bins and the bins where **all** templates have zero entries but the data is non-zero. After completing these procedures, I have a few follow-up questions:
1. During my first trial of fitting the 6 components to an independent MC sample with the `da` method, I did not set any yield limits, and some of the yields went negative. Is this expected?   
<img src="https://user-images.githubusercontent.com/60195377/226501673-290eb352-e3fa-4e93-82c4-d67c79d699f7.png" alt="Failed trial" width="40%">
2. In my second trial, the yield limits were set to [0,∞) and the fit converged.
<img src="https://user-images.githubusercontent.com/60195377/226502439-cd12c12c-040f-4bf5-9590-ec786b00e70b.png" alt="Better trial" width="40%">
However, some yields are different from the expected values, especially in the $D\tau\nu$ component, where the deviation is very large. (The true yield column is calculated from MC truth.) My understanding of this problem is that some components have very similar shapes and the fitter cannot distinguish between them, as indicated by the high correlations. (we also notice the correlations are high). Does this make sense, and if so, do you have any suggestions?

|Component|True Yield|Fitted Yield|Hesse Error|
|:---|:---|:---|:---|
|$D\tau\nu$|8,728|2.3e3|1.0e3|
|$D^\ast\tau\nu$|4,322|4.7e3|1.2e3|
|$D\ell\nu$|144,129|149.3e3|1.2e3|
|$D^\ast\ell\nu$|95,339|93.8e3|1.4e3|
|$D^{\ast\ast}\tau\nu$|2,582|3.8e3|1.0e3|
|$D^{\ast\ast}\ell\nu$|60,091|61.3e3|1.2e3|
<img src="https://user-images.githubusercontent.com/60195377/226505228-df960c37-9512-4442-a337-5c2d5261caad.png" alt="correlation" width="70%">
<img src="https://user-images.githubusercontent.com/60195377/226751351-d0082550-262f-4939-8ee9-789411f3d39d.png" alt="draw_mnmatrix" width="100%">

**Reply by HDembinski** [reply]
Looks like there is a bug in the plotting of the contours, which occurs sometimes (left column). I have an idea how to fix this, but don't know for sure how to reproduce this.

**Reply by HDembinski** [reply]
Before you ask more questions, please send me a minimal example that allows me to reproduce this plot of the contour matrix.

**Reply by zjkjsd** [reply]
Sorry, my bad! I didn't realize that I should upload an example! Now I have attached a template json and a test_data json file in this reply, please read them as follows:
[templates_data.zip](https://github.com/scikit-hep/iminuit/files/11121148/templates_data.zip)
**Edit**: Sorry the test_data.json above seems to not work well, I uploaded the correct one below.
[New.zip](https://github.com/scikit-hep/iminuit/files/11127326/New.zip)

```
# load template
file_path = 'templates.json'
with open(file_path, encoding="utf-8") as file:
    ws = json.load(file)
electron_channel = ws['channels'][0]
sample_names = [s['name'] for s in electron_channel['samples']]
templates = np.array([s['data'] for s in electron_channel['samples']])

# load test_data
file_path2 = 'test_data.json'
with open(file_path2, encoding="utf-8") as file:
    ws2 = json.load(file)
electron_channel2 = ws2['channels'][0]
data = np.array([s['data'] for s in electron_channel2['samples']])
counts=np.sum(data,axis=0).reshape(41,47)

# remove empty bins in the template and test_data
indices_non_zero = np.where(np.sum(templates,axis=0).reshape(41,47) != 0) # mask for removing empty bins in the template
temp = [i.reshape(41,47)[indices_non_zero] for i in templates]
test_data = counts[indices_non_zero]
xedge = np.linspace(0, len(test_data), len(test_data)+1)
init_yields = [3000]*6

# run minuit
c = Template(test_data, xedge, temp, method="da")
m = Minuit(c, *init_yields, name=sample_names)
m.limits = (0, None)
m.simplex().migrad(ncall=1000000)
m.hesse()
```

**Reply by zjkjsd** [reply]
Another warning message I always see with this example is as follows (sometimes it repeats many times): Could you give me some hints why it happens? Thank you!
```
E VariableMetricBuilder Initial matrix not pos.def.
```

**Reply by HDembinski** [reply]
> E VariableMetricBuilder Initial matrix not pos.def.

This can mean a lot of things, none of them good.
1) You provide bad initial values for which the cost function is not positive definite
2) The cost function has numerical issues.

I have no proof, but I expect that the Template cost function is positive definite, so I would rule out 1), also your starting values look ok at first glance. It is possible that you run into numerical issues.

**Reply by zjkjsd** [reply]
I upgraded my iminuit to version 2.21.2 a few days ago and noticed a new behavior. In the example above, the template `temp` is modified (the shape is the same but some elements are different) after running Minuit (see code block below).
```
c = Template(test_data, xedge, temp, method="da")
m = Minuit(c, *init_yields, name=sample_names)
m.limits = (0, None)
m.simplex().migrad(ncall=1000000)
m.hesse()
```
For example, running the minimization for the first time showed a visualization below:
![download](https://user-images.githubusercontent.com/60195377/229324181-1da2cfa3-6de5-4246-b6c9-6601570deb8f.png)
Running the same code block for the second time showed:
![download-1](https://user-images.githubusercontent.com/60195377/229324290-d88386b5-b4cb-47c5-aade-2e6bac3a1042.png)
However, if I reassign the `temp` variable before running the minimization for the second time, the result will be the same as the first time. Is this behavior expected, or is it a bug?

**Reply by HDembinski** [reply]
That looks like a bug, the Template class should not modify the `temp` array.

**Reply by HDembinski** [reply]
I can reproduce this. This must be a bug somewhere in the Template class. This is very strange, because for triggering the bug, it is enough to do this
```py
Template(test_data, xedge, temp, method="da")
c = Template(test_data, xedge, temp, method="da")
m = Minuit(c, *init_yields, name=sample_names)
m.limits = (0, None)
m.simplex().migrad(ncall=1000000)
m.hesse()
```

**Reply by HDembinski** [reply]
The bug is fixed in #856, I will release a new iminuit version with this fix later today. I am sorry that this bug slipped through, despite all the testing that I do.

**Reply by zjkjsd** [reply]
Thank you very much for your help!

**Reply by HDembinski** [reply]
Thank you for providing your data and code, so I could reproduce the problem. If I find the time, I will use this to investigate the numerical issues, perhaps there is a way to make the likelihood more stable using Kahan summation.

**Comment by HDembinski** [comment]
Yes, you need to set limits. If you don't do that, the fit may chose a combination of amplitudes where one or more bins get an expected count that is negative, which is mathematically not possible. It is your responsibility to set the parameter limits so that this never happens.

This tutorial explains the problem in more detail.
https://iminuit.readthedocs.io/en/stable/notebooks/scipy_and_constraints.html

Unless your are in the rase case where you need to fit an interference pattern, the amplitudes of components are always positive.

In a future version of iminuit, limits for the amplitudes may be set automatically by default, so that you have to explicitly unset them if you really want negative amplitudes.

**Reply by zjkjsd** [reply]
Thank you for directing me to the tutorial. I have a follow-up question:
The example in the tutorial doesn't set limits for the signal amplitude but applies a probability density constraint at each data point using scipy. Is it reasonable to follow this procedure for real-world applications where amplitudes of some components are expected to be small? Thank you!

**Reply by HDembinski** [reply]
Yes, sure. The example was made to demonstrate how this can be used in practice.

**Reply by zjkjsd** [reply]
Thank you for your response! I wonder why we should set limits for template fitting but not for the example in the tutorial. Is it because the template fitting is essentially a least-square fitting? (I am not sure if this is true). 

Another question: I want to apply a constraint:
$$\frac{yield(D\tau\nu)}{yield(D^\ast\tau\nu)}=\frac{yield(D\ell\nu)}{yield(D^\ast\ell\nu)}* \frac{efficiency(\tau\nu)}{efficiency(\ell\nu)}$$
The efficiency of $\tau\nu$ and $\ell\nu$ can be calculated from signal MC. Could you please share some insights about how to implement this constraint? Thank you very much for your help!

**Reply by HDembinski** [reply]
In a template fit with two templates, you compute the expected rate `mu` in bin `i` as `mu[i] = a * t1[i] + b * t2[i]`, where `t1` and `t2` are the templates. Mathematically, `mu[i]` must be positive, because the expectation in a bin cannot be negative. `t1` and `t2` are always non-negative in every bin (bins where both t1 and t2 are zero are discarded by the likelihood), because they are derived from histograms. So there are two ways to make `mu[i]` positive everywhere. 1) You use a minimizer in which you can straight-forwardly define the constraint "`mu[i]` must be positive everywhere". Minuit cannot do this, but Scipy can. Note that one of `a` and `b` can be negative in this case, but not both. 2) You apply a stronger constraint, namely that `a` and `b` must be both positive. That is a constraint that Minuit supports.

**Reply by zjkjsd** [reply]
Thank you so much for explaining! It is clearer to me now! I am trying to follow the scipy example in a template fit but am stuck on setting up the constraint. I would appreciate it if you could take a look at it in the code below. And perhaps give me some hints on the yield efficiency constraint in my last question if you have time, thank you very much!
```
from scipy.optimize import NonlinearConstraint
c3 = Template(data, xedge, temp, method="da") # 'jsc', 'asy'
m3 = Minuit(c3, *init_yields, name=samples_name)
m3.limits = (None, None)
m3.simplex().scipy(constraints=NonlinearConstraint(lambda m3: c3.prediction(m3.values)[0], 0, np.inf))
m3.hesse()
```

**Reply by HDembinski** [reply]
The function in the constraint must accept the same number of arguments as your cost function. This should work:
`m3.simplex().scipy(constraints=NonlinearConstraint(lambda *args: c3.prediction(*args)[0], 0, np.inf))`

**Reply by zjkjsd** [reply]
Thank you again! This got my scipy trial to work! There is a small typo in the code above, it should be:
`m3.simplex().scipy(constraints=NonlinearConstraint(lambda *args: c3.prediction(args)[0], 0, np.inf))`.
Note the `prediction(args)`. For users who come across this question in the future, this is because the c3.prediction() only takes a `Sequence[float]` argument.

**Comment by HDembinski** [comment]
Off-topic: @zjkjsd You may like this PR https://github.com/scikit-hep/iminuit/pull/858
